Add ecs_attribute module.

anryko opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
ecs_attribute

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.0 (ecs_attribute 0088c74894) last updated 2017/01/25 01:21:30 (GMT +200)
```

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
In ecs_service (https://github.com/ansible/ansible/pull/20503) I added functionality to constrain container placement based on attributes (as one of placement_constraints expression options). This module will provide possibility to assign attributes to ecs container instances based on ec2 instance id.
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

<!-- Paste verbatim command output below, e.g. before and after your change -->
```
# Set attribute
- ecs_attribute:
    cluster: test-cluster
    name: flavor
    value: test
    state: present
    ec2_instance_id: "{{ ec2_id }}"
  delegate_to: localhost

# Delete attribute
- ecs_attribute:
    cluster: test-cluster
    name: flavor
    state: absent
    ec2_instance_id: "{{ ec2_id }}"
  delegate_to: localhost
```

anryko committed:
ecs_attribute: new module.

ecs_attribute: fixes for willthames and wimnat review 1.

ecs_attribute: fixed docs.

ecs_attribute: fixed docs 2.

ecs_attribute: changed state to 'present' by default.

ecs_attribute: pep8 fix.

ecs_attribute: improved attribute iteration.

ecs_attribute: fixed absent EC2_REGION error message.

ecs_attribute: fixed exception handling on invalid attribute (e.g. 2+2).

ecs_attribute: pep8 fix.

ecs_attribute: docs and python3 fixes.

ecs_attribute: fixed class declaration for python 2-3 compatibility.

gundalow commented on lib/ansible/modules/cloud/amazon/ecs_attribute.py:
```diff
@@ -0,0 +1,264 @@
+options:
+    cluster:
+        description:
+            - The short name or full Amazon Resource Name (ARN) of the cluster that contains the resource to apply attributes. If you do not specify a cluster, the default cluster is assumed
```
All `descriptions:` should be full sentences, and therefore end with a full stop.

gundalow reviewed (changes requested):
needs_revision

Please add this to `CHANGELOG.md` in the root of the repo under the new modules in 2.3 section

gundalow commented:
Please also add an update to `CHANELLOG.md` in this PR

anryko commented:
@gundalow what is ```CHANELLOG.md```?

simplesteph commented:
@anryko I'll review this on Monday. In the meantime I believe ansible prefers for you not to have "devel merges" into your commits and do a rebase instead if need be. Referring to commit b5aab93

gundalow commented:
Hi, you can ignore the `CHANGELOG.md` request.

anryko commented:
@gundalow removed ```CHANGELOG.md``` commit.
@simplesteph reverted merge.

simplesteph commented on lib/ansible/modules/cloud/amazon/ecs_attribute.py:
```diff
@@ -0,0 +1,264 @@
+            ec2_instances = self.ecs.describe_container_instances(cluster=self.cluster,
+                    containerInstances=ecs_instances_arns)['containerInstances']
+        except botocore.exceptions.ClientError as e:
+            self.module.fail_json(msg="Can't connet to the cluster - %s" % str(e))
```
typo here `connet`

simplesteph commented on lib/ansible/modules/cloud/amazon/ecs_attribute.py:
```diff
@@ -0,0 +1,264 @@
+        except StopIteration as e:
+            self.module.fail_json(msg="EC2 instance Id not found in ECS cluster" % (str(e)))
+
+        return ecs_arn['containerInstanceArn']
```
just for naming clarity, can we have `ecs_arn = next(i for i in ec2_instances if i['ec2InstanceId'] == ec2_id)['containerInstanceArn']` ? or rename `ecs_arn`

simplesteph commented on lib/ansible/modules/cloud/amazon/ecs_attribute.py:
```diff
@@ -0,0 +1,264 @@
+
+        attr_obj = {
+                'targetType': 'container-instance',
+                'targetId': self._ec2_id_to_ecs_arn(ec2_id),
```
so I haven't tried the API but just wondering if we could do things more efficiently. 
http://boto3.readthedocs.io/en/latest/reference/services/ecs.html#ECS.Client.put_attributes 

For `targetId` it says that if you use `targetType` as `container-instance` you can use the short form ID for a resource instead. Therefore is that doable to do `targetId: ec2_id` ?

anryko replied:
You can use only ecs container instance id here. If I try to pass ec2 instance id I get ```An error occurred (InvalidParameterException) when calling the PutAttributes operation: instanceId shorter than 36.``` For my purposes I want to use ec2 id and have no use for ecs id, this is the reason for this ugly conversion. However if someone uses attributes in a different way and would prefer assigning attribute based on ecs container-instance id, this code should be straightforward  to extend.

simplesteph commented on lib/ansible/modules/cloud/amazon/ecs_attribute.py:
```diff
@@ -0,0 +1,264 @@
+            matched_ecs_targets = self.ecs.list_attributes(
+                    cluster=self.cluster, **attr_obj)['attributes']
+        except botocore.exceptions.ClientError as e:
+            self.module.fail_json(msg="Can't connet to the cluster - %s" % str(e))
```
same typo here

simplesteph reviewed (changes requested):
Overall looks good, just wondering about if the API calls could be simplified. 

Also, to put things in perspective, can you give an example of how this would be chained to other modules in a sample playbook?

anryko commented:
@simplesteph 
OK, I'll try to give an example, it will not be a fully working one, because it's all done in roles with abstracted configuration in my case, so hard to extract only needed parts and edit them to work standalone. Nevertheless it will show general idea of the execution flow...

Firs I would create an ecs cluster:
```
- name: ecs instances
  hosts: localhost
  connection: local
  gather_facts: False
  tasks:
    - name: create new cluster
      ecs_cluster:
        name: test-cluster
        state: present
```
Then find ami and create container-nodes
```
    - name: find latest ami
      ec2_ami_find:
        name: 'amzn-ami-*-amazon-ecs-optimized'
        owner: 591542846629
        sort: name
        sort_order: descending
        sort_end: 1
      register: ami_find

    - name: setup ec2 instances
      ec2:
        image: "{{ ami_find.results[0].ami_id }}"
        instance_tags: test_cluster
        ...
        user_data: |
          #!/bin/bash
          cat > /etc/ecs/ecs.config <<- EOF
          ECS_CLUSTER=test-cluster
          ECS_ENGINE_AUTH_TYPE=dockercfg
          ECS_ENGINE_AUTH_DATA={{ dockerhub_authentications | to_json }}
          ECS_ENGINE_TASK_CLEANUP_WAIT_DURATION=10m
          EOF
      register: ec2

    - name: wait for ssh
      wait_for:
        host: "{{ item.tagged_instances[0].private_ip }}"
        search_regex: "OpenSSH"
        port: 22
        timeout: 300
      with_items:
        - "{{ ec2.results }}"
```
Then we need to refresh ec2.py cache to pick up newly created ec2 instances
```
    - meta: refresh_inventory
```
Now I can put attributes on the instances and do system setup in the second part of the playbook, based on the tag I added while creating ec2 instances
```
- name: common stuff
  hosts: tag_test_cluster
  become: True
  gather_facts: True
  tasks:
    - name: put attributes
      become: False
      ecs_attribute:
        cluster: test-cluster
        name: flavor
        value: test
        state: present
        ec2_instance_id: "{{ ec2_id }}"
     delegate_to: localhost
```

And proceed setting up timezone, monitoring, centralized logging, etc.

mattclay commented:
CI failure due to PEP 8 issues:

```
2017-01-31 18:51:03 ERROR: PEP 8: lib/ansible/modules/cloud/amazon/ecs_attribute.py:46:161: E501 line too long (189 > 160 characters) (current)
2017-01-31 18:51:03 ERROR: PEP 8: lib/ansible/modules/cloud/amazon/ecs_attribute.py:157:17: E126 continuation line over-indented for hanging indent (current)
2017-01-31 18:51:03 ERROR: PEP 8: lib/ansible/modules/cloud/amazon/ecs_attribute.py:180:17: E126 continuation line over-indented for hanging indent (current)
2017-01-31 18:51:03 ERROR: PEP 8: lib/ansible/modules/cloud/amazon/ecs_attribute.py:186:21: E126 continuation line over-indented for hanging indent (current)
2017-01-31 18:51:03 ERROR: PEP 8: lib/ansible/modules/cloud/amazon/ecs_attribute.py:231:13: E126 continuation line over-indented for hanging indent (current)
2017-01-31 18:51:03 ERROR: PEP 8: lib/ansible/modules/cloud/amazon/ecs_attribute.py:239:9: E121 continuation line under-indented for hanging indent (current)
```

simplesteph commented:
This looks good to me. I haven't figured if this will cover all use cases but people will open / issues and PR is that happen. Thanks for work @anryko 

shipit

anryko commented:
@simplesteph @gundalow ready

simplesteph commented:
@gundalow it still shows you want a review. Can you please fix ? 
Shipit

gundalow reviewed (approved):
shipit

willthames commented on lib/ansible/modules/cloud/amazon/ecs_attribute.py:
```diff
@@ -0,0 +1,251 @@
+
+try:
+    import boto
+    import botocore
```
`import botocore` should be in the HAS_BOTO3 check.

I'm not convinced that `boto` is required here - you don't seem to use it at all - can you remove all references to HAS_BOTO if that's the case.

anryko replied:
You are right. Will fix that.

wimnat commented:
This module doesn't appear idempotent. 

Shouldn't you be able to specify all the attributes at once as a list? What happens to exists attributes that are not specified? Should there be a purge option? Maybe the ec2_tag module can provide some ideas

anryko commented:
@wimnat it is minimal in functionality, nevertheless it's idempotent. Attributes that are not specified are not changed. You can set attribute and remove attribute by name on a specific ec2 instance. Multiple attributes can be managed using ```with_items```.

wimnat commented:
@anryko I disagree. You are not enforcing a state you are just proceding to add or or remove an attribute.

anryko commented:
@wimnat we can discuss definition of 'state', or you could provide an example of ansible task that you think would better describe the action. Does this look better in your opinion?
```yaml
- ecs_attribute:
    state: present
    cluster: test-cluster
    ec2_instance_id: "{{ ec2_id }}"
    attributes:
      - flavor: test
      - migrated
```

willthames commented:
@anryko I think your last suggestion is closer to what @wimnat has in mind (as he did mention the `ec2_tag` module which behaves similarly)

Of course that brings a different problem of whether you then need a `purge` option along the lines of the `ec2_group` module.

The issue is of course what happens on the second module call when you add a separate attribute - should it replace the first attribute. `ec2_tag` just replaces tags if they're not present in the list on the second call, `ec2_group` only removes rules not present if the `purge_rules` option is set.

anryko commented:
@willthames @wimnat 
- removed boto dependency
- changed ansible task api to support multiple attributes like in proposed example

```purge_rules``` equivalent is problematic because aws has presetup attributes like:
```
      "name": "com.amazonaws.ecs.capability.task-iam-role",
      "name": "com.amazonaws.ecs.capability.task-iam-role-network-host",
      "name": "ecs.ami-id",
      "name": "ecs.availability-zone",
      "name": "ecs.instance-type",
      "name": "ecs.os-type"
```
Those can be filtered out by ```ecs.*|com.amazonaws.*```, but I would prefer not to go there at the moment. Initial task syntax is very important, so I do appreciate @wimnat comments, however all not essential functionality can be added later as a reaction to community needs. All I needed from this module so far is to add one custom attribute for container constrains.

mattclay commented on lib/ansible/modules/cloud/amazon/ecs_attribute.py:
```diff
@@ -0,0 +1,290 @@
+                required: true
+            value:
+                description:
+                    - The value of the attribute. Up to 128 letters (uppercase and lowercase), numbers, hyphens, underscores, periods, at signs (@), forward slashes, colons, and spaces are allowed.
```
CI failure due to PEP 8 issue:

```
2017-02-14 22:18:30 ERROR: PEP 8: lib/ansible/modules/cloud/amazon/ecs_attribute.py:52:161: E501 line too long (197 > 160 characters) (current)
```

The PEP 8 tests can be run locally with `make pep8`.

anryko replied:
@mattclay thanks for the tip!

willthames commented on lib/ansible/modules/cloud/amazon/ecs_attribute.py:
```diff
@@ -0,0 +1,297 @@
+
+    def __bool__(self):
+        return bool(self.attributes)
+    __nonzero__=__bool__
```
This should have a blank line before and spaces around the assignment to meet PEP8

anryko replied:
right, sorry, copy paste... :)

s-hertel commented on lib/ansible/modules/cloud/amazon/ecs_attribute.py:
```diff
@@ -0,0 +1,310 @@
+
+    @staticmethod
+    def _validate_attrs(attrs):
+        return all(('name', 'value') == tuple(attr.keys()) for attr in attrs)
```
This isn't really working right for me since the dict keys are not necessarily in that order. Need some sorting.
Running this task:
```
      ecs_attribute:
        cluster: "{{ name }}"
        state: absent
        attributes:
          - flavor: test
          - migrated
        ec2_instance_id: "{{ ec2_id }}"
```
Results in:
```
fatal: [localhost]: FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_args": {
            "attributes": [
                {
                    "flavor": "test"
                },
                "migrated"
            ],
            "aws_access_key": null,
            "aws_secret_key": null,
            "cluster": "testcluster",
            "ec2_instance_id": "i-07da5958fc74fbc",
            "ec2_url": null,
            "profile": "shertel",
            "region": "us-east-1",
            "security_token": null,
            "state": "absent",
            "validate_certs": true
        }
    },
    "msg": "Incorrect attribute format - {'value': 'test', 'name': 'flavor'}"
}

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1
```
I think this is happening because [line 255 ](https://github.com/ansible/ansible/pull/20618/files#diff-bf8ee38d299d949473d0a2221a344250R255) is calling EcsAttributes for the second time, after those attributes were already parsed once. So this error doesn't seem to happen when initially creating attributes, only when updating or removing already existing ones.

anryko replied:
Couldn't reproduce on my setup.
Added both possibilities to the validation check. Works faster and more explicit than sorting.

s-hertel replied:
Fixed the problem, thanks.

s-hertel commented on lib/ansible/modules/cloud/amazon/ecs_attribute.py:
```diff
@@ -0,0 +1,310 @@
+            if isinstance(attr, dict):
+                if len(attr) != 1:
+                    self.module.fail_json(msg="Incorrect attribute format - %s" % str(attr))
+                name, value = attr.items()[0]
```
should be list(attr.items())[0] to work with python3 too.
```
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_xs5l34yw/ansible_module_ecs_attribute.py", line 310, in <module>
    main()
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_xs5l34yw/ansible_module_ecs_attribute.py", line 280, in main
    attrs = EcsAttributes(module, attributes)
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_xs5l34yw/ansible_module_ecs_attribute.py", line 131, in __init__
    self.attributes = attributes if self._validate_attrs(attributes) else self._parse_attrs(attributes)
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_xs5l34yw/ansible_module_ecs_attribute.py", line 151, in _parse_attrs
    name, value = attr.items()[0]
TypeError: 'dict_items' object does not support indexing

fatal: [localhost]: FAILED! => {
    "changed": false,
    "failed": true,
    "module_stderr": "Traceback (most recent call last):\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_xs5l34yw/ansible_module_ecs_attribute.py\", line 310, in <module>\n    main()\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_xs5l34yw/ansible_module_ecs_attribute.py\", line 280, in main\n    attrs = EcsAttributes(module, attributes)\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_xs5l34yw/ansible_module_ecs_attribute.py\", line 131, in __init__\n    self.attributes = attributes if self._validate_attrs(attributes) else self._parse_attrs(attributes)\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_xs5l34yw/ansible_module_ecs_attribute.py\", line 151, in _parse_attrs\n    name, value = attr.items()[0]\nTypeError: 'dict_items' object does not support indexing\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 0
}

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1
```

anryko replied:
Thanks! Fixed.
Module now works with python3.

anryko commented:
@s-hertel thanks! Will look in to it in the next couple of days.

anryko commented:
@willthames ping.

ryansb commented on lib/ansible/modules/cloud/amazon/ecs_attribute.py:
```diff
@@ -0,0 +1,311 @@
+from ansible.module_utils.ec2 import boto3_conn, ec2_argument_spec, get_aws_connection_info
+
+
+class EcsAttributes:
```
Need to use new-style Python 2 classes, as this Py3 class declaration won't behave right in Python 2 environments. Change to `class EcsAttributes(object):`

ryansb commented on lib/ansible/modules/cloud/amazon/ecs_attribute.py:
```diff
@@ -0,0 +1,311 @@
+        return EcsAttributes(self.module, attrs_diff)
+
+
+class Ec2EcsInstance:
```
Change to `class Ec2EcsInstance(object):`

anryko commented:
@ryansb thanks! done.

anryko commented:
@ryansb, @willthames: is there anything else i can do to get this merged?

s-hertel commented:
Reran shippable, all green.

s-hertel commented:
Thanks for working on this @anryko! Thanks everyone else for the reviews. :)
